providers/implementations/keymgmt/ec_kmgmt.c

```diff
@@ -0,0 +1,253 @@
+        goto err;
+
+    if ((ret = EC_KEY_new()) == NULL
+            || !EC_KEY_set_group(ret, ecg)
+            || !EC_KEY_set_private_key(ret, priv_key)
+            || !EC_KEY_set_public_key(ret, pub_point))
+        goto err;
+
+    ok = 1;
+
+ err:
+    BN_clear_free(priv_key);
+    OPENSSL_free(pub_key);
+    EC_POINT_free(pub_point);
+    EC_GROUP_free(ecg);
+    if (!ok) {
+        EC_KEY_free(ret);
+        ret = NULL;
+    }
+    return ret;
+}
+
+static ossl_inline
+int key_to_params(const EC_KEY *eckey, OSSL_PARAM_BLD *tmpl)
+{
+    const BIGNUM *priv_key = NULL;
+    const EC_POINT *pub_point = NULL;
+    const EC_GROUP *ecg = NULL;
+    char *pub_key = NULL;
+    int ret = 0;
+
+    if (eckey == NULL)
+        return 0;
+
+    ecg = EC_KEY_get0_group(eckey);
+    priv_key = EC_KEY_get0_private_key(eckey);
+    pub_point = EC_KEY_get0_public_key(eckey);
+
+    /* group and public_key must be present, priv_key is optional */
+    if (ecg == NULL || pub_point == NULL)
+        return 0;
+#define PNT_FORM POINT_CONVERSION_UNCOMPRESSED
+    if ((pub_key = EC_POINT_point2hex(ecg, pub_point, PNT_FORM, NULL)) == NULL)
+        return 0;
+
+    if (!domparams_to_params(ecg, tmpl))
+        goto err;
+    if (priv_key != NULL
+            && !ossl_param_bld_push_BN(tmpl,
+                                       OSSL_PKEY_PARAM_EC_PRIV_KEY,
+                                       priv_key))
+        goto err;
+    if (!ossl_param_bld_push_utf8_string(tmpl,
+                                         OSSL_PKEY_PARAM_EC_PUB_KEY,
+                                         pub_key,
+                                         0))
+        goto err;
+
+    ret = 1;
+
+ err:
+    OPENSSL_free(pub_key);
+    return ret;
+}
+
+static
+void *ec_importdomparams(void *provctx, const OSSL_PARAM params[])
+{
+    return params_to_domparams(params);
+}
+
+static
+int ec_exportdomparams(void *domparams, OSSL_CALLBACK *param_cb, void *cbarg)
+{
+    EC_GROUP *ecg = domparams;
+    OSSL_PARAM_BLD tmpl;
+    OSSL_PARAM *params = NULL;
+    int ret;
+
+    ossl_param_bld_init(&tmpl);
+    if (ecg == NULL
+        || !domparams_to_params(ecg, &tmpl)
+        || (params = ossl_param_bld_to_param(&tmpl)) == NULL)
+        return 0;
+    ret = param_cb(params, cbarg);
+    ossl_param_bld_free(params);
+    return ret;
+}
+
+static
+void *ec_importkey(void *provctx, const OSSL_PARAM params[])
+{
+    return params_to_key(params);
+}
+
+static
+int ec_exportkey(void *key, OSSL_CALLBACK *param_cb, void *cbarg)
+{
+    EC_KEY *eck = key;
+    OSSL_PARAM_BLD tmpl;
+    OSSL_PARAM *params = NULL;
+    int ret;
+
+    ossl_param_bld_init(&tmpl);
+    if (eck == NULL
+        || !key_to_params(eck, &tmpl)
+        || (params = ossl_param_bld_to_param(&tmpl)) == NULL)
+        return 0;
+    ret = param_cb(params, cbarg);
+    ossl_param_bld_free(params);
+    return ret;
+}
+
+const OSSL_DISPATCH ec_keymgmt_functions[] = {
+    /*
```

Comment:
Probaby should get rid of this TODO: as EXPORTKEY already exists.

Reply:
Good point :)

Copy pasted from `dh_keymgmt.c` :)

Reply:
Done!